### DISCUSSION TITLE
Listeners for Firestore?

### DISCUSSION THREAD

**Discussion by PadreMontoya** [GodotNuts/GodotFirebase#186] [Q&A] [answered]
It looks like there is support for listeners with the Realtime Database, but not Firestore, correct?  Just making sure I'm not overlooking something simple.  I looked at the source code and most of it seems HTTP based.

Would it be implemented similar to how it was for the Realtime Database?  Debating on whether to take a shot at it.

**Comment by BearDooks** [comment]
@PadreMontoya apologies it took us so long to get back to you. 

What are you trying to do? That might make a difference with what part of Firebase you want to use.

Also yes a lot of the code is HTTP based since we are using the RestAPI for Firebase.

~ Chuck

**Comment by WolfgangSenff** [comment, marked as answer]
It is not implemented the same way as the Realtime Database, unfortunately, as Firestore does not support dynamic updates through REST. It only supports them through gRPC, which Godot itself does not support, which makes me very sad.

Thanks for the question though! Sorry it took us so long, as @BearDooks said.

**Comment by PadreMontoya** [comment]
Ok, no problem at all - thanks for getting back to me!  I'm chugging
forward with the Realtime Database, though it isn't as shiny as Firestore.
:)   Thanks for your work on the project - it's a lifesaver!

>

**Reply by WolfgangSenff** [reply]
It's not as shiny perhaps, but there's an awesome little thing for it that makes it really clean, which is a Firebase Extension (that's what they're called) that can limit the number of child nodes at a path. This can make it so if you don't care about the game history, you can funnel moves to it, and they'll be interpreted and discarded as new moves are sent. It's really nice to prevent your DB from overflowing with data for fast-moving games that require a lot of moves, or have a lot of players doing moves at once. I'll be demonstrating this in my talk at GodotCon this year, which will be on more generic cloud-based backend patterns for games rather than Firebase, but should have some really useful information none-the-less. :)

**Comment by PadreMontoya** [comment]
That brings up something I'm wondering - is Firebase practical for fast
moving games?

I'm doing a relatively simple Blackjack-style game, so I don't imagine
there will be more than 100 server-bound posts per 5 minutes per game.
(Players have 20 seconds to decide whether to "Stay" (Hit) or "Flee"
(Fold), then the server resolves the round once everyone has decided)

I have it functioning and it works, I'm just worried about being crushed by
a Firebase bill.

Storage should be trivial, but should I be concerned about traffic?

**Reply by WolfgangSenff** [reply]
No - for physics-based games or FPS or anything that is meant to be real-time, despite the name of the database, it is not suitable for those types of games. Check out my talk at GodotCon this year for some interesting discussions/patterns regarding how to use any cloud-based solutions (it's not about Firebase specifically, but rather the patterns to use in order to make games with it).

**Reply by WolfgangSenff** [reply]
Regarding the Firebase bill, it's pretty tough to say what you might experience with that type of game. I still haven't actually had any bill come up because the free tier is so good. Actually, after GodotCon, I may get one, because I'm using Firestore in a game that I'll be posting around the time of the convention and announcing, but other than that, I'm reasonably sure it will be almost free, despite even using an Extension to fix swear words.

**Reply by WolfgangSenff** [reply]
Finally, regarding storage, it can be expensive for sure. I don't know what, generally, that's going to cost you if you're using it directly. If you're talking about just storing stuff in Firestore and the Realtime Database, that's pretty cheap, but Storage is not, in my experience.
